common-npm-packages/artifacts-common/EntraWifUserServiceConnectionUtils.ts

```diff
@@ -2,76 +2,96 @@ import path = require("path");
 import * as tl from 'azure-pipelines-task-lib/task';
 import { getSystemAccessToken } from "./webapi";
 import fetch from "node-fetch";
+import { retryOnException } from "./retryUtils";
 
 tl.setResourcePath(path.join(__dirname, 'module.json'), true);
 
 const ADO_RESOURCE : string = "499b84ac-1321-427f-aa17-267ca6975798/.default";
 const CLIENT_ASSERTION_TYPE : string = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
 const GRANT_TYPE = "client_credentials";
+const PPE_HOSTS : string [] = ["vsts.me","codedev.ms","devppe.azure.com"];
 
-export async function getFederatedWorkloadIdentityCredentials(serviceConnectionName: string, tenantId?: string) : Promise<string | undefined>
-{
-    try {
-        let tenant = tenantId ?? tl.getEndpointAuthorizationParameterRequired(serviceConnectionName, "TenantId");
-        tl.debug(tl.loc('Info_UsingTenantId', tenantId));
-        const systemAccessToken = getSystemAccessToken();
-        const url = process.env["SYSTEM_OIDCREQUESTURI"]+"?api-version=7.1&serviceConnectionId="+serviceConnectionName;
-        
-        const ADOResponse: {oidcToken: string} = await (await fetch(url, 
-        {
+export async function getFederatedWorkloadIdentityCredentials(serviceConnectionName: string, tenantId?: string) : Promise<string | undefined>{
+    let tenant = tenantId ?? tl.getEndpointAuthorizationParameterRequired(serviceConnectionName, "TenantId");
+    tl.debug(tl.loc('Info_UsingTenantId', tenantId));
+    const systemAccessToken = getSystemAccessToken();
+    const url = process.env["SYSTEM_OIDCREQUESTURI"]+"?api-version=7.1&serviceConnectionId="+serviceConnectionName;
+    
+    return await retryOnException(async () => {
+        var oidcToken = await fetch(url, {
             method: 'POST', 
-            headers: 
-            {
+            headers: {
                 'Content-Type': 'application/json',
                 'Authorization': 'Bearer '+ systemAccessToken
             }
-        })).json() as {oidcToken: string};
+        }).then(async response => {
+            var oidcObject = await (response?.json()) as {oidcToken: string};
 
-        tl.setSecret(ADOResponse.oidcToken);
-        let entraURI = "https://login.windows.net/"+tenant+"/oauth2/v2.0/token";
+            if (!oidcObject?.oidcToken){
```

Comment:
Out of curiosity, what is oidc? It looks like it's part of the flow to gather tokens but curious to hear more about it

Reply:
OIDC (short for openID Connect) is a protocol built on top of the Oauth protocol. It allows the federated credential flow. More info here: https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols. I can provide more resources offline too.